atomic_reactor/source.py

```diff
@@ -35,21 +35,24 @@
 class SourceConfig(object):
     """ read container.yaml file from build source and store as attrs """
 
-    def __init__(self, build_path):
-        self.data = {}
+    def __init__(self, build_path, provider_params=None):
+        git_params = provider_params or {}
+        # validated in input_osv3 if its here at all
+        self.data = git_params.get('container_yaml', {})
         self.file_path = os.path.join(build_path, REPO_CONTAINER_CONFIG)
-        if os.path.exists(self.file_path):
-            try:
-                # read file and validate against schema
-                self.data = read_yaml_from_file_path(
-                    self.file_path, 'schemas/container.json'
-                ) or {}
-            except Exception:
-                logger.exception(
-                    "Failed to load and validate source config YAML from %s",
-                    self.file_path
-                )
-                raise
+        if not self.data:
+            if os.path.exists(self.file_path):
+                try:
+                    # read file and validate against schema
+                    self.data = read_yaml_from_file_path(
+                        self.file_path, 'schemas/container.json'
```

Comment:
Instead of this can we use `jsonschema.RefResolver()` to automatically load the right external schema when we initially validate the user_params object, i.e. in `read_yaml`?

The reason this isn't just a one-liner is that we load schema from the package resources rather than simply reading files from the filesystem, but I think we make it do that as well without too much bother, something like:
```
pkgres_scheme = 'pkgresource://'
def load_resource_from_uri(uri):
    resource = resource_stream('atomic_reactor', uri[len(pkgres_scheme):])
    schema = codecs.getreader('utf-8')(resource)
    return json.load(schema)

resolver = jsonschema.RefResolver('pkgresource://', schema,
    handlers={'pkgresource', load_resource_from_url})
...
validator.validate(..., resolver=resolver)
...
```

Reply:
I'm running into this issue: https://github.com/Julian/jsonschema/issues/343

Here's the current code:
```
def read_yaml(yaml_data, schema):
    """
    :param yaml_data: string, yaml content
    """

    pkgres_scheme = 'pkgresource://'
    def load_resource_from_url(uri):
        resource = resource_stream('atomic_reactor/schemas/', uri[len(pkgres_scheme):])
        schema = codecs.getreader('utf-8')(resource)
        return json.load(schema)

    ar_resolver = jsonschema.RefResolver(pkgres_scheme, schema,
        handlers={'pkgresource': load_resource_from_url})

    try:
        resource = resource_stream('atomic_reactor', schema)
        schema = codecs.getreader('utf-8')(resource)
    except (IOError, TypeError):
        logger.error('unable to extract JSON schema, cannot validate')
        raise

    try:
        schema = json.load(schema)
    except ValueError:
        logger.error('unable to decode JSON schema, cannot validate')
        raise
    data = yaml.safe_load(yaml_data)
    validator = jsonschema.Draft4Validator(schema=schema, resolver=ar_resolver)
    try:
        jsonschema.Draft4Validator.check_schema(schema)
        validator.validate(data)
```

its' failing on the internal references in plugins.json.  if I fix those, with this:
```
  "properties": {
    "prebuild_plugins": {"$ref": "#/definitions/general_plugins_phase"},
    "buildstep_plugins": {"$ref": "#/definitions/base_plugins_phase"},
    "postbuild_plugins": {"$ref": "pkgresource://plugins.json#/definitions/general_plugins_phase"},
    "prepublish_plugins": {"$ref": "#/definitions/general_plugins_phase"},
    "exit_plugins": {"$ref": "#/definitions/general_plugins_phase"}
  }
```
then I get this error
```
atomic_reactor/plugins/input_osv3.py:208: in run
    read_yaml(json.dumps(self.plugins_json), 'schemas/plugins.json')
atomic_reactor/util.py:1371: in read_yaml
    validator.validate(data)
/usr/lib/python2.7/site-packages/jsonschema/validators.py:116: in validate
    for error in self.iter_errors(*args, **kwargs):
/usr/lib/python2.7/site-packages/jsonschema/validators.py:95: in iter_errors
    for error in errors:
/usr/lib/python2.7/site-packages/jsonschema/_validators.py:285: in properties_draft4
    schema_path=property,
/usr/lib/python2.7/site-packages/jsonschema/validators.py:108: in descend
    for error in self.iter_errors(instance, schema):
/usr/lib/python2.7/site-packages/jsonschema/validators.py:95: in iter_errors
    for error in errors:
/usr/lib/python2.7/site-packages/jsonschema/_validators.py:198: in ref
    with validator.resolver.resolving(ref) as resolved:
/usr/lib64/python2.7/contextlib.py:17: in __enter__
    return self.gen.next()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jsonschema.validators.RefResolver object at 0x7f8694a3e8d0>
ref = 'pkgresource://plugins.json#/definitions/general_plugins_phase'

    @contextlib.contextmanager
    def resolving(self, ref):
        """
            Context manager which resolves a JSON ``ref`` and enters the
            resolution scope of this ref.
    
            :argument str ref: reference to resolve
    
            """
    
        full_uri = urljoin(self.resolution_scope, ref)
        uri, fragment = urldefrag(full_uri)
        if not uri:
            uri = self.base_uri
    
        if uri in self.store:
            document = self.store[uri]
        else:
            try:
                document = self.resolve_remote(uri)
            except Exception as exc:
>               raise RefResolutionError(exc)
E               RefResolutionError: Import by filename is not supported.

/usr/lib/python2.7/site-packages/jsonschema/validators.py:296: RefResolutionError
```
at which point I am stuck and confused.

Reply:
I decided to resolve this by not passing container.yaml.  Shallow clones are fast enough to hopefully make this a non-issue.